src/config/mod.rs

```diff
@@ -27,165 +27,172 @@ pub(crate) mod file_lines;
 #[allow(unreachable_pub)]
 pub(crate) mod lists;
 pub(crate) mod macro_names;
+pub(crate) mod style_edition;
 
 // This macro defines configuration options used in rustfmt. Each option
 // is defined as follows:
 //
 // `name: value type, default value, is stable, description;`
 create_config! {
     // Fundamental stuff
-    max_width: usize, 100, true, "Maximum width of each line";
-    hard_tabs: bool, false, true, "Use tab characters for indentation, spaces for alignment";
-    tab_spaces: usize, 4, true, "Number of spaces per tab";
-    newline_style: NewlineStyle, NewlineStyle::Auto, true, "Unix or Windows line endings";
-    indent_style: IndentStyle, IndentStyle::Block, false, "How do we indent expressions or items";
+    max_width: MaxWidth, true, "Maximum width of each line";
+    hard_tabs: HardTabs, true, "Use tab characters for indentation, spaces for alignment";
+    tab_spaces: TabSpaces, true, "Number of spaces per tab";
+    newline_style: NewlineStyleConfig, true, "Unix or Windows line endings";
+    indent_style: IndentStyleConfig, false, "How do we indent expressions or items";
```

Comment:
Not a change we need to make now (or ever), but as I was working on this I wondered if we could generalize `StyleEditionDefault` and rename it to something like `RustfmtConfigOption`. That trait could also be used to define each configuration options description as well as its default value for different style editions.

For example

```rust
use crate::config::StyleEdition;
use crate::config::config_type::ConfigType;

pub trait RustfmtConfigOption {
    type ConfigType: ConfigType;

    fn default(style_edition: StyleEdition) -> Self::ConfigType;
    fn description() -> &'static str;
}
```

That way, `create_config!` could focus on setting field names, ty's that impl `RustfmtConfigOption`, and the stability of each option.

Reply:
I like the sound of this, but agree it's something worth deferring